### PR TITLE
Add a new rule to detect integer overflow when converting between integer types

### LINKED ISSUE #1130: Add detection of overflow during integer conversion
### Summary 

G109 provides detection of overflow during strconv.Atoi, but there seems to be no good way to detect the more common conversions between integers, such as uint32 to int32.

### PR DESCRIPTION
**Pull request by ccojocar**
fixes #1130

### CODE REVIEW

**Comment by remyleone** [comment]
Could you add documentation about how to fix it and make the linter pass? 🙏

File: analyzers/conversion_overflow.go
Comment (ccoVeille) [inline analyzers/conversion_overflow.go:54]: I might have missed something about the implementation, but I would like to suggest this


```suggestion
							fmt.Sprintf("possible integer overflow conversion %s -> %s", src, dst),
```

It might not happen depending on people code.

It's better to report a possible error, than affirming there is one. People would argue.

File: analyzers/conversion_overflow.go
Comment (ccoVeille) [inline analyzers/conversion_overflow.go:79]: This regexp will be computed each time parseInt will be called, it should be moved out the function

File: analyzers/conversion_overflow.go
Comment (ccoVeille) [inline analyzers/conversion_overflow.go:86]: Please add something like this

```suggestion
	// int64 => int 64, uint8 => uint 8
	it := matches[re.SubexpIndex("type")]
	is := matches[re.SubexpIndex("size")]
```

File: analyzers/conversion_overflow.go
Comment (ccoVeille) [inline analyzers/conversion_overflow.go:130]: This 8 is magical number here

File: analyzers/conversion_overflow.go
Comment (ccoVeille) [inline analyzers/conversion_overflow.go:106]: Maybe this

```suggestion
func couldIntOverflow(src string, dst string) bool {
```

File: testutils/g115_samples.go
Comment (ccoVeille) [inline testutils/g115_samples.go:5]: I'm a bit worried these tests are only focusing on things that could be reported.

I would expect to be added to cover conversion of things that should pass.

They are likely the ones to report false positive if something is invalid in the current implementation, or if a later refactoring screw everything

**Review by ccoVeille — CHANGES_REQUESTED** [review]

**Comment by FairlySadPanda** [comment]
This check has rolled out and ergo suddenly comments on it!

This check seems over-prescriptive. In the builtin functions, len returns type int, despite only returning zero or positive ints. Casting out of len into uint32 is thus more declarative. This is just one example where this change enforces a //nolint flag.

The act of casting is an assertion that the programmer knows what they are doing.

**Comment by ccojocar** [comment]
> There is perfectly valid code that can be guaranteed not to overflow that this check catches. For example;

The type of v is still `int` which on a 64 bit system is 64 bits. Essentially the rule sees a conversion from int64 to int32. It is not so smart to detect the conversion done before using `strconv.ParseInt`.  It think adding these extension can be useful but for now you can make the size explicit by selecting the `int32` type for v instead of `int`. This will make the size explicit into the type.

See more discussions in #1185.

Also if this rule doesn't work for your use case, feel free to exclude it but in some cases integer  overflow can lead to security issues and there isn't a protection for this in the go runtime at the moment. 

```
gosec -exclude G115 ./...
```

**Comment by moitias** [comment]
> @moitias The type of v is still `int` which on a 64 bit system is 64 bits.

Well, if we're being pedantic, no, the type of `v` is not `int` but `int64` because ParseInt and ParseUInt return explicit int64s/uint64s but yeah, this is just nitpicking.

> It think adding these extension can be useful but for now you can make the size explicit by selecting the `int32` type for v instead of `int`. This will make the size explicit into the type.

Without commit access to Go stdlib (and considering Go 1 compatibility promise), I don't think I can do that. 

> Essentially the rule sees a conversion from int64 to int32. It is not so smart to detect the conversion done before using `strconv.ParseInt`.

And that, indeed, is the problem. There _are_ safe ways of doing the conversion, but this checker does not allow for them either. 

> Also if this rule doesn't work for your use case, feel free to exclude it but in some cases integer overflow can lead to security issues and there isn't a protection for this in the go runtime at the moment.

Yeah, but this checker does not allow the developer to do it safely either, so they will need to disable the checker (locally or globally), making this checker essentially useless for any case that actually needs to do conversions and of course, nothing but lost CPU time for any case that doesn't do conversions. So, I'd argue that the usefulness of this check is dependent on it being 'smart'. If it is not, people will disable it and overflows might then creep in.

**Comment by ccojocar** [comment]
@moitias I am looking forward to your contributions to improve this rule. I think is more productive to do this instead of just complaining and using an OSS tool maintain by volunteers ;-).

**Comment by moitias** [comment]
Look, I appreciate the work you are putting into this (and helping me out) but I'm trying to point out - to help _you_ out in turn - that this rule in its current form seems to be a bit prematurely turned on because it seems to have no use cases at all. This is because if one has it enabled, there is no way of doing the conversion securely while in fact there obviously are. 

Could you please, to clarify the reasoning for this rule, at least describe one way of doing a conversion to a smaller bit size safely or are you of the opinion that it is never acceptable/secure to do? I would hope we could at least agree that linter warnings should have a way of fixing them, right?

**Comment by ccoVeille** [comment]
> @moitias @FairlySadPanda @ccoVeille I am looking forward to your contributions to improve this rule. I think is more productive to do this instead of just complaining and using an OSS tool maintained by volunteers ;-). 

I can understand your reaction. I could say _Touché_. But I'm more likely to reply you positively than being affected. I'm a smiling person.

Be sure I have nothing against you or the idea you got. I wish the issues we are facing now where found before it was merge, but there were not.

You can understand people are reacting when their CI is now a Christmas tree 🎄

I wish I had reviewed your PR earlier. I wish others may have reviewed earlier.
Someone may have raised the consequences of this PR.

There problems was not with the idea or the code, but with the review of your changes suggestions.

That said, sometimes shit happens. Let's try to find a solution.

**Comment by co60ca** [comment]
I came to see if this issue had be reported yet.

But we are getting
G115: integer overflow conversion uint8 -> int (gosec)
and according to the spec uint8 always fits in an int without overflow.

File: analyzers/conversion_overflow.go
Comment (ccoVeille) [reply analyzers/conversion_overflow.go:54]: @ccojocar what do you think about the changes I suggested?

Do you agree with them?

Would you change it? Or do you want me to open a PR for them?

**Comment by ccoVeille** [comment]
For anyone interested, the discussion continued here

https://github.com/securego/gosec/issues/1185

**Pull request merged**
